toolbox.py

```diff
@@ -1,33 +1,55 @@
 #! /usr/bin/python3
 """### Provides various small functions for the average workflow."""
 
-__all__ = ['loop2d', 'loop3d', 'writeBook', 'placeLectern',
-           'placeInventoryBlock', 'placeSign', 'getOptimalDirection']
-__version__ = 'v4.2_dev'
+__all__ = ['isSequence', 'normaliseCoordinates', 'loop2d', 'loop3d',
+           'writeBook', 'placeLectern', 'placeInventoryBlock', 'placeSign',
+           'getOptimalDirection', 'visualizeHeightmap']
+__version__ = 'v4.3_dev'
 __year__ = '2021'
 __author__ = 'Blinkenlights'
 
 from functools import lru_cache
+from itertools import product
 from random import choice
 
+import cv2
 import lookup
-from interfaceUtils import getBlock
-from interfaceUtils import globalinterface as gi
-from interfaceUtils import runCommand
+import numpy as np
+from interface import getBlock
+from interface import globalinterface as gi
+from interface import runCommand
+from matplotlib import pyplot as plt
 
 
-def loop2d(x1, y1, x2=None, y2=None):
+def isSequence(sequence):
+    """**Determine whether sequence is a sequence**."""
+    try:
+        sequence[0:-1]
+        return True
+    except TypeError:
+        return False
+
+
+def normaliseCoordinates(*args):
+    """**Return set of coordinates where (x1, y1, z1) <= (x2, y2, z2)**."""
+    args1, args2 = args[:len(args) // 2], args[len(args) // 2:]
+    return [a if a < b else b for a, b in zip(args1, args2)] \
+        + [a if a > b else b for a, b in zip(args1, args2)]
+
+
+def loop2d(a1, b1, a2=None, b2=None):
     """**Return all coordinates in a 2D region**.
 
-    If only one pair is provided, the loop will yield x1*z1 values
+    If only one pair is provided, the loop will yield a1*b1 values
     If two pairs are provided the loop will yield all results between them
         inclusively
     """
-    if x2 is None or y2 is None:
-        x1, y1, x2, y2 = 0, 0, x1 - 1, y1 - 1
-    for x in range(x1, x2 + 1):
-        for y in range(y1, y2 + 1):
-            yield x, y
+    if a2 is None or b2 is None:
+        a1, b1, a2, b2 = 0, 0, a1 - 1, b1 - 1
+
+    a1, b1, _, a2, b2, _ = normaliseCoordinates(a1, b1, 0, a2, b2, 0)
```

Comment:
A question might be whether normaliseCoordinates should
a) be Americanized to `normalizeCoordinates`
b) be able to handle D2 coordinates separately, as many other functions do